File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/NimbusBuilder.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.experiments.nimbus

import android.content.Context
import android.content.SharedPreferences
import android.net.Uri
import androidx.annotation.RawRes
import kotlinx.coroutines.runBlocking
import org.mozilla.experiments.nimbus.internal.FeatureManifestInterface

private const val TIME_OUT_LOADING_EXPERIMENT_FROM_DISK_MS = 200L

/**
 * A builder for [Nimbus] singleton objects, parameterized in a declarative class.
 */
abstract class AbstractNimbusBuilder<T : NimbusInterface>(val context: Context) {

    /**
     * An optional server URL string.
     *
     * This will only be null or empty in development or testing, or in any build variant of a
     * non-Mozilla fork.
     */
    var url: String? = null

    /**
     * A closure for reporting errors from Rust.
     */
    var errorReporter: ErrorReporter = { _: String, _: Throwable -> }

    /**
     * A flag to select the main or preview collection of remote settings. Defaults to `false`.
     */
    var usePreviewCollection: Boolean = false

    /**
     * A flag to indicate if this is being run on the first run of the app. This is used to control
     * whether the `initial_experiments` file is used to populate Nimbus.
     */
    var isFirstRun: Boolean = true

    /**
     * A optional raw resource of a file downloaded at or near build time from Remote Settings.
     */
    @RawRes
    var initialExperiments: Int? = null

    /**
     * The timeout used to wait for the loading of the `initial_experiments`.
     */
    var timeoutLoadingExperiment: Long = TIME_OUT_LOADING_EXPERIMENT_FROM_DISK_MS

    /**
     * Optional callback to be called after the creation of the nimbus object and it is ready
     * to be used.
     */
    var onCreateCallback: (T) -> Unit = {}

    /**
     * Optional callback to be called everytime experiments have been applied.
     *
     * Experiment recipes are usually fetched shortly after startup, and those pending recipes are
     * applied at the following startup.
     *
     * This is not usually needed.
     */
    var onApplyCallback: () -> Unit = {}

    /**
     * Optional callback to be called everytime experiments have been fetched.
     *
     * Experiment recipes are usually fetched shortly after startup, and those pending recipes are
     * applied at the following startup.
     *
     * This is not usually needed.
     */
    var onFetchCallback: () -> Unit = {}

    /**
     * The `object` generated from the `nimbus.fml.yaml` file and the nimbus-gradle-plugin.
     */
    var featureManifest: FeatureManifestInterface<*>? = null

    /**
     * The shared preferences used to configure the app.
     */
    var sharedPreferences: SharedPreferences? = null

    /**
     * Build a [Nimbus] singleton for the given [NimbusAppInfo]. Instances built with this method
     * have been initialized, and are ready for use by the app.
     *
     * Instance have _not_ yet had [fetchExperiments()] called on it, or anything usage of the
     * network. This is to allow the networking stack to be initialized after this method is called
     * and the networking stack to be involved in experiments.
     */
    fun build(appInfo: NimbusAppInfo): T {
        // Eventually we'll want to use `NimbusDisabled` when we have no NIMBUS_ENDPOINT.
        // but we keep this here to not mix feature flags and how we configure Nimbus.
        val serverSettings: NimbusServerSettings? = if (!url.isNullOrBlank()) {
            if (usePreviewCollection) {
                NimbusServerSettings(url = Uri.parse(url), collection = "nimbus-preview")
            } else {
                NimbusServerSettings(url = Uri.parse(url))
            }
        } else {
            null
        }

        // Is the app being built locally, and the nimbus-cli
        // hasn't been used before this run.
        fun NimbusInterface.isLocalBuild() = url.isNullOrBlank() && isFetchEnabled()

        @Suppress("TooGenericExceptionCaught")
        return try {
            newNimbus(appInfo, serverSettings).apply {
                // Apply any experiment recipes we downloaded last time, or
                // if this is the first time, we load the ones bundled in the res/raw
                // directory.
                val job = if (initialExperiments != null && (isFirstRun || isLocalBuild())) {
                    applyLocalExperiments(initialExperiments!!)
                } else {
                    applyPendingExperiments()
                }

                // We always want initialize Nimbus to happen ASAP and before any features (engine/UI)
                // have been initialized. For that reason, we use runBlocking here to avoid
                // inconsistency in the experiments.
                // We can safely do this because Nimbus does most of its work on background threads,
                // including the loading the initial experiments from disk. For this reason, we have a
                // `joinOrTimeout` to limit the blocking until `timeoutLoadingExperiment`.
                runBlocking {
                    // We only read from disk when loading first-run experiments. This is the only time
                    // that we should join and block. Otherwise, we don't want to wait.
                    job.joinOrTimeout(timeoutLoadingExperiment)
                }
                // By now, on this thread, we have a fully initialized Nimbus object, ready for use:
                // * we gave a 200ms timeout to the loading of a file from res/raw
                // * on completion or cancellation, applyPendingExperiments or initialize was
                //   called, and this thread waited for that to complete.
                featureManifest?.initialize { this }
                onCreateCallback(this)
            }
        } catch (e: Throwable) {
            // Something went wrong. We'd like not to, but stability of the app is more important than
            // failing fast here.
            errorReporter("Failed to initialize Nimbus", e)
            newNimbusDisabled()
        }
    }

    /**
     * Construct a new [NimbusInterface] object with the passed parameters.
     */
    protected abstract fun newNimbus(
        appInfo: NimbusAppInfo,
        serverSettings: NimbusServerSettings?,
    ): T

    /**
     * In the event of the error constructing or configuring a Rust backed
     * [NimbusInterface] object, then construct a dummy object.
     */
    protected abstract fun newNimbusDisabled(): T

    /**
     * Creates the [NimbusDeviceInfo] for each Nimbus object built.
     */
    protected open fun createDeviceInfo() = NimbusDeviceInfo.default()

    /**
     * Creates the [NimbusDelegate] for each Nimbus instance built.
     *
     * The delegate is the main low-level interface between the embedding apps and the SDK.
     *
     * Override this if you want to customize the threading, error reporting or logging used
     * by the [NimbusInterface] instance.
     */
    protected open fun createDelegate() = NimbusDelegate.default()

    /**
     * Creates the observer used to tie together the feature manifest and the SDK. Implementers of
     * [newNimbus] should use this to register this observer with the [NimbusInterface] object.
     */
    protected fun createObserver(): NimbusInterface.Observer =
        Observer(featureManifest, onFetchCallback, onApplyCallback)

    /**
     * Returns a list of feature ids that support coenrollment. Implementers of [newNimbus] should
     * use this to pass into the [NimbusInterface] instance.
     */
    protected fun getCoenrollingFeatureIds(): List<String> =
        featureManifest?.getCoenrollingFeatureIds() ?: listOf()
}

private class Observer(
    val featureManifest: FeatureManifestInterface<*>?,
    val onFetchCallback: () -> Unit,
    val onApplyCallback: () -> Unit,
) : NimbusInterface.Observer {
    override fun onExperimentsFetched() {
        onFetchCallback.invoke()
    }

    override fun onUpdatesApplied(updated: List<EnrolledExperiment>) {
        featureManifest?.invalidateCachedValues()
        onApplyCallback.invoke()
    }
}

class DefaultNimbusBuilder(context: Context) : AbstractNimbusBuilder<NimbusInterface>(context) {
    override fun newNimbus(appInfo: NimbusAppInfo, serverSettings: NimbusServerSettings?) =
        Nimbus(
            context,
            appInfo = appInfo,
            prefs = sharedPreferences,
            coenrollingFeatureIds = getCoenrollingFeatureIds(),
            server = serverSettings,
            deviceInfo = createDeviceInfo(),
            delegate = createDelegate(),
            observer = createObserver(),
        )

    override fun newNimbusDisabled() = NullNimbus(context)
}

File: components/nimbus/ios/Nimbus/NimbusBuilder.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation

/**
 * A builder for [Nimbus] singleton objects, parameterized in a declarative class.
 */
public class NimbusBuilder {
    let dbFilePath: String

    public init(dbPath: String) {
        dbFilePath = dbPath
    }

    /**
     * An optional server URL string.
     *
     * This will only be null or empty in development or testing, or in any build variant of a
     * non-Mozilla fork.
     */
    @discardableResult
    public func with(url: String?) -> Self {
        self.url = url
        return self
    }

    var url: String?

    /**
     * A closure for reporting errors from Rust.
     */
    @discardableResult
    public func with(errorReporter reporter: @escaping NimbusErrorReporter) -> NimbusBuilder {
        errorReporter = reporter
        return self
    }

    var errorReporter: NimbusErrorReporter = defaultErrorReporter

    /**
     * A flag to select the main or preview collection of remote settings. Defaults to `false`.
     */
    @discardableResult
    public func using(previewCollection flag: Bool) -> NimbusBuilder {
        usePreviewCollection = flag
        return self
    }

    var usePreviewCollection: Bool = false

    /**
     * A flag to indicate if this is being run on the first run of the app. This is used to control
     * whether the `initial_experiments` file is used to populate Nimbus.
     */
    @discardableResult
    public func isFirstRun(_ flag: Bool) -> NimbusBuilder {
        isFirstRun = flag
        return self
    }

    var isFirstRun: Bool = true

    /**
     * A optional raw resource of a file downloaded at or near build time from Remote Settings.
     */
    @discardableResult
    public func with(initialExperiments fileURL: URL?) -> NimbusBuilder {
        initialExperiments = fileURL
        return self
    }

    var initialExperiments: URL?

    /**
     * The timeout used to wait for the loading of the `initial_experiments
     */
    @discardableResult
    public func with(timeoutForLoadingInitialExperiments seconds: TimeInterval) -> NimbusBuilder {
        timeoutLoadingExperiment = seconds
        return self
    }

    var timeoutLoadingExperiment: TimeInterval = 0.200 /* seconds */

    /**
     * Optional callback to be called after the creation of the nimbus object and it is ready
     * to be used.
     */
    @discardableResult
    public func onCreate(callback: @escaping (NimbusInterface) -> Void) -> NimbusBuilder {
        onCreateCallback = callback
        return self
    }

    var onCreateCallback: ((NimbusInterface) -> Void)?

    /**
     * Optional callback to be called after the calculatoin of new enrollments and applying of changes to
     * experiments recipes.
     */
    @discardableResult
    public func onApply(callback: @escaping (NimbusInterface) -> Void) -> NimbusBuilder {
        onApplyCallback = callback
        return self
    }

    var onApplyCallback: ((NimbusInterface) -> Void)?

    /**
     * Resource bundles used to look up bundled text and images. Defaults to `[Bundle.main]`.
     */
    @discardableResult
    public func with(bundles: [Bundle]) -> NimbusBuilder {
        resourceBundles = bundles
        return self
    }

    var resourceBundles: [Bundle] = [.main]

    /**
     * The object generated from the `nimbus.fml.yaml` file and the nimbus-gradle-plugin.
     */
    @discardableResult
    public func with(featureManifest: FeatureManifestInterface) -> NimbusBuilder {
        self.featureManifest = featureManifest
        return self
    }

    var featureManifest: FeatureManifestInterface?

    /**
     * The command line arguments for the app. This is useful for QA, and can be safely left in the app in production.
     */
    @discardableResult
    public func with(commandLineArgs: [String]) -> NimbusBuilder {
        self.commandLineArgs = commandLineArgs
        return self
    }

    var commandLineArgs: [String]?

    /**
     * Build a [Nimbus] singleton for the given [NimbusAppSettings]. Instances built with this method
     * have been initialized, and are ready for use by the app.
     *
     * Instance have _not_ yet had [fetchExperiments()] called on it, or anything usage of the
     * network. This is to allow the networking stack to be initialized after this method is called
     * and the networking stack to be involved in experiments.
     */
    public func build(appInfo: NimbusAppSettings) -> NimbusInterface {
        let serverSettings: NimbusServerSettings?
        if let string = url,
           let url = URL(string: string)
        {
            if usePreviewCollection {
                serverSettings = NimbusServerSettings(url: url, collection: remoteSettingsPreviewCollection)
            } else {
                serverSettings = NimbusServerSettings(url: url, collection: remoteSettingsCollection)
            }
        } else {
            serverSettings = nil
        }

        do {
            let nimbus = try newNimbus(appInfo, serverSettings: serverSettings)
            let fm = featureManifest
            let onApplyCallback = onApplyCallback
            if fm != nil || onApplyCallback != nil {
                NotificationCenter.default.addObserver(forName: .nimbusExperimentsApplied,
                                                       object: nil,
                                                       queue: nil)
                { _ in
                    fm?.invalidateCachedValues()
                    onApplyCallback?(nimbus)
                }
            }
            if let args = unpack(args: commandLineArgs), let experiments = args["experiments"] {
                // If we have command line arguments, then load experiments from there,
                // and disable future fetching.
                nimbus.setExperimentsLocally(experiments)
                nimbus.applyPendingExperiments().waitUntilFinished()
                // setExperimentsLocally and applyPendingExperiments run on the
                // same single threaded dispatch queue, so we can run them in series,
                // and wait for the apply.
                nimbus.setFetchEnabled(false)
            } else if let file = initialExperiments, isFirstRun || serverSettings == nil {
                let job = nimbus.applyLocalExperiments(fileURL: file)
                _ = job.joinOrTimeout(timeout: timeoutLoadingExperiment)
            } else {
                nimbus.applyPendingExperiments().waitUntilFinished()
            }

            // By now, on this thread, we have a fully initialized Nimbus object, ready for use:
            // * we gave a 200ms timeout to the loading of a file from res/raw
            // * on completion or cancellation, applyPendingExperiments or initialize was
            //   called, and this thread waited for that to complete.
            featureManifest?.initialize { nimbus }
            onCreateCallback?(nimbus)

            return nimbus
        } catch {
            errorReporter(error)
            return newNimbusDisabled()
        }
    }

    private func unpack(args: [String]?) -> [String: String]? {
        guard let args = args else {
            return nil
        }
        if !args.contains("--nimbus-cli") {
            return nil
        }

        var argMap = [String: String]()
        var key: String?
        args.forEach { arg in
            var value: String?
            switch arg {
            case "--version":
                key = "version"
            case "--experiments":
                key = "experiments"
            default:
                value = arg
            }

            if let k = key, let v = value {
                argMap[k] = v
                key = nil
                value = nil
            }
        }

        if argMap["version"] != "1" {
            return nil
        }

        guard let experiments = argMap["experiments"],
              let payload = try? Dictionary.parse(jsonString: experiments),
              payload["data"] is [Any]
        else {
            return nil
        }

        return argMap
    }

    func newNimbus(_ appInfo: NimbusAppSettings, serverSettings: NimbusServerSettings?) throws -> NimbusInterface {
        try Nimbus.create(serverSettings, appSettings: appInfo, dbPath: dbFilePath,
                          resourceBundles: resourceBundles, errorReporter: errorReporter)
    }

    func newNimbusDisabled() -> NimbusInterface {
        NimbusDisabled.shared
    }
}
